backend/arm64/isa/arm64_ast.ml

```diff
@@ -0,0 +1,597 @@
+end
+
+module Instruction = struct
+  type t =
+    { name : Instruction_name.t;
+      operands : Operand.t array
+    }
+
+  let create name operands = { name; operands }
+
+  let print ppf t =
+    let { name; operands } = t in
+    let pp_sep = Operand.print_separator in
+    Format.fprintf ppf "%s\t%a"
+      (Instruction_name.to_string name)
+      (Format.pp_print_seq ~pp_sep Operand.print)
+      (Array.to_seq operands)
+end
+
+module Asm = struct
+  (** Emit assembly instructions for gas. *)
+
+  (* CR-soon gyorsh: The directives defined below duplicate [X86_ast]. Another
+     version of these directives is in [backend/asm_targets]. Refactor and
+     remove duplication. *)
+
+  type constant =
+    | Const of int64
+    | ConstThis
+    | ConstLabel of string
+    | ConstLabelOffset of string * int
+    | ConstAdd of constant * constant
+    | ConstSub of constant * constant
+
+  type line =
+    | Ins of Instruction.t
+    | Align of bool * int
+    | Byte of constant
+    | Bytes of string
+    | Comment of string
+    | Global of string
+    | Protected of string
+    | Hidden of string
+    | Weak of string
+    | Long of constant
+    | NewLabel of string
+    | NewLine
+    | Quad of constant
+    | Section of string list * string option * string list * bool
+    | Sleb128 of constant
+    | Space of int
+    | Uleb128 of constant
+    | Word of constant
+    (* gas only (the masm emitter will fail on them) *)
+    | Cfi_adjust_cfa_offset of int
+    | Cfi_endproc
+    | Cfi_startproc
+    | Cfi_remember_state
+    | Cfi_restore_state
+    | Cfi_def_cfa_register of string
+    | Cfi_def_cfa_offset of int
+    | File of int * string (* (file_num, file_name) *)
+    | Indirect_symbol of string
+    | Loc of
+        { file_num : int;
+          line : int;
+          col : int;
+          discriminator : int option
+        }
+    | Private_extern of string
+    | Set of string * constant
+    | Size of string * constant
+    | Type of string * string
+    (* MacOS only *)
+    | Direct_assignment of string * constant
+
+  type prog = line list
+
+  let print_line ppf line =
+    match line with
+    | Ins i -> Format.fprintf ppf "\t%a\n" Instruction.print i
+    | NewLine | Cfi_endproc | Cfi_startproc | Cfi_remember_state
+    | Cfi_restore_state | Byte _ | Bytes _ | Comment _ | Global _ | Protected _
+    | Hidden _ | Weak _ | Long _ | NewLabel _ | Quad _
+    | Section (_, _, _, _)
+    | Sleb128 _ | Space _ | Uleb128 _ | Word _ | Cfi_adjust_cfa_offset _
+    | Cfi_def_cfa_register _ | Cfi_def_cfa_offset _
+    | File (_, _)
+    | Indirect_symbol _ | Loc _ | Private_extern _
+    | Set (_, _)
+    | Size (_, _)
+    | Type (_, _)
+    | Direct_assignment (_, _)
+    | Align _ ->
+      Misc.fatal_error "Directive not implemented yet"
+
+  let print_program ppf lines = List.iter (print_line ppf) lines
+
+  let generate_asm oc lines =
+    let ppf = Format.formatter_of_out_channel oc in
+    print_program ppf lines
```

Comment:
I might be midly worried `Format` could
insert line break - is that a valid concern?

Reply:
I don't know but it doesn't look like it, I'm not emitting any break hints or boxes, and the assembler will complain. Would you `Printf` instead throuhgout this module? Using `Format` in this mordule because it has more helper functions, but it's easy to change. In any case `generate_asm` is not used at the moment, the entire `Asm.Acc` is here to get rid of `.mlp`.

Reply:
Ah, yes, sorry, since there are no boxes
that should be fine indeed.